### DISCUSSION TITLE
macOS build process

### DISCUSSION THREAD

**Discussion by AndersGoran** [jamulussoftware/jamulus#3321] [Technical Support & Help] [answered]
This might be a bit of a catch-22 if there are really no active Mac devs in this project. 

I saw the shoutout from Chris Rimple in the Worldjam group on Facebook, looking for Mac developers. I'm a Mac & Jamulus user so I thought I should see if I could compile and run from source. I follow the Mac build instructions, and I do get a Jamulus.app but it crashes because I think it can't find the Qt stuff from Homebrew. The resulting binary is just 4MB. 

If I use the `mac/deploy_mac.sh` script, I get a much larger binary that does work, but that script takes such a long time to run in comparison with just `xcodebuild build`. I can tell that the deploy script bundles a lot of Qt stuff into the Jamulus.app bundle. 

**Is there anybody here that sees what I'm missing?** 

Do I really need to go through the deploy process to get something that runs? 

The crash report contains this: 
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x1886e2d38 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x188717ee0 pthread_kill + 288
2   libsystem_c.dylib             	       0x188652330 abort + 168
3   QtCore                        	       0x10423aa14 0x104220000 + 109076
4   QtCore                        	       0x1044fa264 QMessageLogger::fatal(char const*, ...) const + 116
5   QtGui                         	       0x103cbc334 QGuiApplicationPrivate::createPlatformIntegration() + 7420
6   QtGui                         	       0x103cbc358 QGuiApplicationPrivate::createEventDispatcher() + 36
7   QtCore                        	       0x104402ea8 QCoreApplicationPrivate::init() + 1164
8   QtGui                         	       0x103cb6684 QGuiApplicationPrivate::init() + 68
9   QtWidgets                     	       0x103181738 QApplicationPrivate::init() + 36
10  QtWidgets                     	       0x103181878 QApplication::QApplication(int&, char**, int) + 164
11  Jamulus                       	       0x102545bb0 main + 21272 (main.cpp:823)
12  dyld                          	       0x102b3d08c start + 520
```

and if I try starting the app from the CLI, with QT_DEBUG=1, it says:

```
*[main][~/Developer/jamulus]$ ./Release/Jamulus.app/Contents/MacOS/Jamulus
- allocated port number: 22134
QFactoryLoader::QFactoryLoader() checking directory path "/opt/homebrew/Cellar/qt@5/5.15.13_1/plugins" ...
QFactoryLoader::QFactoryLoader() checking directory path "/Users/ango/Developer/jamulus/Release/Jamulus.app/Contents/MacOS" ...
QFactoryLoader::QFactoryLoader() looking at "/Users/ango/Developer/jamulus/Release/Jamulus.app/Contents/MacOS/Jamulus"
QMachOParser: '/Users/ango/Developer/jamulus/Release/Jamulus.app/Contents/MacOS/Jamulus' is not a valid Mach-O binary (not a dynamic library)
"'/Users/ango/Developer/jamulus/Release/Jamulus.app/Contents/MacOS/Jamulus' is not a valid Mach-O binary (not a dynamic library)"
         not a plugin
QFactoryLoader::QFactoryLoader() checking directory path "/opt/homebrew/Cellar/qt@5/5.15.13_1/plugins/platforms" ...
QFactoryLoader::QFactoryLoader() checking directory path "/Users/ango/Developer/jamulus/Release/Jamulus.app/Contents/MacOS/platforms" ...
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in "/opt/homebrew/Cellar/qt@5/5.15.13_1/plugins"
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

[1]    78021 abort      ./Release/Jamulus.app/Contents/MacOS/Jamulus
```

I do have a libqcocoa here:

```*[main][~/Developer/jamulus]$ file /opt/homebrew/Cellar/qt@5/5.15.13_1/plugins/platforms/libqcocoa.dylib
/opt/homebrew/Cellar/qt@5/5.15.13_1/plugins/platforms/libqcocoa.dylib: Mach-O 64-bit dynamically linked shared library arm64```

**Comment by chrisrimple** [comment]
I believe that @softins is a Mac developer and is on this project, so may be able to debug.

**Reply by softins** [reply, marked as answer]
I do have a Mac, running an old macOS (Catalina 10.15), and have compiled Jamulus on it, but not using the Homebrew stuff. But I am in no way a knowledgable Mac developer.

Looking back through my shell history, and also going from memory, I _think_ I did something like this:

- installed xcode from the appstore
- installed the xcode command line tools using the shell command `xcode-select --install`
- installed Qt to `/usr/local/opt/qt/5.15.2` using the Mac installer from the Qt website.
- created `/etc/paths.d/Qt` containing the line `/usr/local/opt/qt/5.15.2/clang_64/bin` (and then start a new shell session for it to take effect)

Then I was able to do the same steps I was used to doing on Linux:

```
qmake Jamulus.pro
make clean
make
```

After which the Jamulus binary was in `Jamulus.app/Contents/MacOS/Jamulus`

I found it useful to put an executable file called `Jamulus` in the project root, containing:

```
#!/bin/sh
exec Jamulus.app/Contents/MacOS/Jamulus "$@"
```

I could then do what I was used to on Linux: `./Jamulus` followed by any options such as `-s`.

A simple symbolic link did not work, as then the executable didn't find all the required components.

I can't guarantee there are no errors in the above, but I hope it gets you started!

**Reply by AndersGoran** [reply]
This seems to help. If I do both the qmake/make commands to build, and use that sh script to launch the executable, it actually runs. And that's still with the Homebrew Qt, no need to touch anything in `/etc/paths.d`. 
I think I also need to set QT_QPA_PLATFORM_PLUGIN_PATH to some carefully chosen directory under /opt/homebrew
I will do some more digging to understand this better, but at least it runs now without the whole deploy script. 

Thanks!

**Comment by ann0see** [comment]
Thanks for trying. Have a look at https://github.com/jamulussoftware/jamulus/issues/3078

**Reply by AndersGoran** [reply]
Thank you, will look closer - very likely there is something useful in that discussion.

**Reply by ann0see** [reply]
Just as bump, the xcodebuild unfortunately really seems to need the macdeployqt stuff run.
